### ISSUE TITLE
regression in classification of traffic on non-default port

### ISSUE THREAD

**Issue by afiaux** [ntop/nDPI#344]
Hello,

I've noticed that commit 424fbb710ca7a883b7826eebc876c5c0611a3b2a introduced a regression in protocol classification when the traffic is not on the default port.

For instance SSH on port 22 is correctly classified as **SSH**, but the same SSH traffic on port 80 is classified as **Unknown**.

Reverting only the following change from the commit fix the problem:
```diff
diff --git a/src/lib/ndpi_main.c b/src/lib/ndpi_main.c
index 7dd3b04..dd1c7bc 100644
--- a/src/lib/ndpi_main.c
+++ b/src/lib/ndpi_main.c
@@ -3215,7 +3215,8 @@ void check_ndpi_tcp_flow_func(struct ndpi_detection_module_struct *ndpi_struct,
          func = ndpi_struct->proto_defaults[flow->guessed_protocol_id].func;
     }
 
-    if(flow->detected_protocol_stack[0] == NDPI_PROTOCOL_UNKNOWN) {
+    if((flow->detected_protocol_stack[0] == NDPI_PROTOCOL_UNKNOWN)
+       && (flow->guessed_protocol_id == NDPI_PROTOCOL_UNKNOWN)) {
       for(a = 0; a < ndpi_struct->callback_buffer_size_tcp_payload; a++) {
         if((func != ndpi_struct->callback_buffer_tcp_payload[a].func)
           && (ndpi_struct->callback_buffer_tcp_payload[a].ndpi_selection_bitmask & *ndpi_selection_packet) == ndpi_struct->callback_buffer_tcp_payload[a].ndpi_selection_bitmask

```
However I'm not sure if it's really the cause of the problem or just a side effect.

Here are two PCAPs with the same SSH traffic, once on port 22 and once on port 80: [ndpi-ssh.zip](https://github.com/ntop/nDPI/files/788523/ndpi-ssh.zip)

Thank you in advance for your help!

**Comment by kYroL01** [comment]
Thanks! I'll check asap

**Comment by kYroL01** [comment]
I just start to see the problem but I think, on a fast look, that the check added is necessary. I'll go deeper and let u know.
Thanks for the pcap

**Comment by afiaux** [comment]
Hello,

So I've looked a bit at the code and I really don't get what you are trying to accomplish with the protocol guessing.

Whether it's in `ndpi_l4_detection_process_packet` or in `ndpi_detection_process_packet`:
- you first call `ndpi_guess_protocol_id` which will set `flow->guessed_protocol_id` to the protocol associated to the default port (80 -> HTTP, 22 -> SSH, ...)
- The you call `check_ndpi_flow_func` which will do the following:
  - if `flow->guessed_protocol_id` is set, only check for this protocol
  - if `flow->guessed_protocol_id` is not set, check for all protocols

So basically, for the traffic on a port **X** which is considered as a default port for protocol **Y**, only protocol **Y** is checked.

So in the end you either need to:
- not guess the protocol until the end, or just let the user decide to call `ndpi_guess_undetected_protocol` if the classification is **Unknown**.
- not limit the checked protocol to `flow->guessed_protocol_id` in the `check_ndpi_*_flow_func` functions (which corresponds to the revert mentioned in the bug report).

Kind regards,

**Comment by kYroL01** [comment]
@afiaux I'm agree with you. Must be changed to improve the correct way to detect.

**Comment by lucaderi** [comment]
@kYroL01 Any prediction about when we can expect this fix?

**Comment by kYroL01** [comment]
@lucaderi I'm doing some modification of the code to improve performance and fix the issue. I hope in a couple of days to fix all
